### PR TITLE
CDAP-1928 Interface for the Condition node in the Workflow.

### PR DESCRIPTION
**Pull request by sagarkapare**
Jira: https://issues.cask.co/browse/CDAP-1928

This PR contains the implementation of the interface for Condition node in the Workflow. 
Following example illustrates the API for the condition:
```java
@Override
public void configure() {
   setName("PurchaseHistoryWorkflow");
   setDescription("PurchaseHistoryWorkflow description");
   addMapReduce("PurchaseHistoryBuilder");
   condition(new MyPredicate())
      .addMapReduce("UserProfileBuilder")
   .otherwise()
      .addMapReduce("CustomBuilder")
   .end();
}
```

where `MyPredicate` is the public class which implements the `Predicate` interface as
```java
public static class MyPredicate implements Predicate<Map<String, String>> {
    @Override
    public boolean apply(@Nullable Map<String, String> input) {
      if (input != null && input.containsKey("someKey")) {
        return true;
      }
      return false;
    }
}
```

If the predicate evaluates to true MapReduce program `UserProfileBuilder` will be executed, otherwise MapReduce program `CustomBuilder` will be executed.

Build is running here: http://builds.cask.co/browse/CDAP-DUT1402-1

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/Predicate.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * @param <T> the type of the input based on which the predicate is to be evaluated
+ */
+public interface Predicate<T> {
+  /**
+   * Returns the result of applying this predicate to {@code input}.
+   * @param input the input based on which the predicate to be evaluated
+   * @return the result of the evaluation of the predicate
+   */
+  boolean apply(@Nullable T input);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/Predicate.java]: Can it be null?

File: cdap-api/src/main/java/co/cask/cdap/api/Predicate.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+  /**
+   * Indicates whether another object is equal to this predicate.
+   * @param object the object to be compared with this predicate
+   * @return the result of the comparison
+   */
+  boolean equals(@Nullable Object object);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/Predicate.java]: I would say we don't need this for now.

**Comment by bdmogal** [comment]
Why don't we call these methods:
```
if(new MyPredicate())
  .addMapReduce(...)
else()
  .addSpark(...)
end()
```
?

File: cdap-api/src/main/java/co/cask/cdap/api/Predicate.java (hunk 2 above)
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/api/Predicate.java]: Ok. will remove it.

File: cdap-api/src/main/java/co/cask/cdap/api/Predicate.java (hunk 1 above)
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/api/Predicate.java]: I made the interface similar to the guava's `Predicate` interface. can remove it.

**Comment by sagarkapare** [comment]
`if`, `else` are keywords in java, we can't use them for the function names.

**Comment by bdmogal** [comment]
oops. stupid question then. ignore :)

File: cdap-api/src/main/java/co/cask/cdap/api/Predicate.java (hunk 1 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/Predicate.java]: hum.. maybe just keep it for now.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowConditionConfigurer.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  implements WorkflowConditionConfigurer<T>, WorkflowConditionAdder {
+
+  private final T parentConfigurer;
+  private final List<WorkflowNode> ifBranch = Lists.newArrayList();
+  private final List<WorkflowNode> elseBranch = Lists.newArrayList();
+  private List<WorkflowNode> currentBranch;
+  private boolean addingToIfBranch = true;
+  private final Predicate<Map<String, String>> predicate;
+
+  public DefaultWorkflowConditionConfigurer(T parentConfigurer, Predicate<Map<String, String>> predicate) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowConditionConfigurer.java]: You only need the predicate class name in the configurer, no need to carry the instance around.

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowConditionNode.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Represents the CONDITION node in the {@link Workflow}.
+ */
+public class WorkflowConditionNode extends WorkflowNode {
+  private final List<WorkflowNode> ifBranch;
+  private final List<WorkflowNode> elseBranch;
+  private final String predicateClassName;
+
+  public WorkflowConditionNode(String nodeId, String predicateClassName, List<WorkflowNode> ifBranch,
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowConditionNode.java]: Should it be public or just package visible?

**Comment by chtyim** [comment]
Just some minor comments. Overall looks ok to me.

File: cdap-api/src/main/java/co/cask/cdap/api/Predicate.java (hunk 2 above)
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/api/Predicate.java]: Removed this method from the interface.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowConditionConfigurer.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowConditionConfigurer.java]: Fixed. Using predicate class name now.

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowConditionNode.java (hunk 4 above)
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowConditionNode.java]: It is used by `DefaultWorkflowConditionConfigurer` which is in package `co.cask.cdap.internal.app.workflow`. Hence it should be public.

**Comment by sagarkapare** [comment]
@chtyim Addressed comments. Please take a look again. Thanks!

**Comment by chtyim** [comment]
LGTM

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-DUT1402-3

**Comment by sagarkapare** [comment]
Build passed. Merging.

**Pull request merged**